ykrt/src/trace/swt/buffer.rs

```diff
@@ -0,0 +1,205 @@
+use std::alloc::{Layout, alloc, dealloc, handle_alloc_error};
+use std::collections::HashMap;
+use yksmp::Location::{Direct, Indirect};
+use yksmp::Record;
+
+use crate::trace::swt::cp::ControlPointStackMapId;
+
+/// A safer wrapper around allocated buffer memory with proper RAII
+#[derive(Debug)]
+pub(crate) struct AlignedBuffer {
+    ptr: *mut u8,
+    layout: Layout,
+}
+
+impl AlignedBuffer {
+    /// Create a new aligned buffer with the specified size
+    pub(crate) unsafe fn new(size: usize) -> Self {
+        let layout = Layout::from_size_align(size, 16)
+            .expect("Failed to create layout for live vars buffer");
+        let ptr = unsafe {
+            let ptr = alloc(layout);
+            if ptr.is_null() {
+                handle_alloc_error(layout);
+            }
+            std::ptr::write_bytes(ptr, 0, size);
+            ptr
+        };
+
+        AlignedBuffer { ptr, layout }
+    }
+
+    /// Get the raw pointer to the buffer
+    pub(crate) fn as_ptr(&self) -> *mut u8 {
+        self.ptr
+    }
+
+    /// Get the size of the buffer
+    pub(crate) fn size(&self) -> usize {
+        self.layout.size()
+    }
+
+    /// Get the layout of the buffer  
+    pub(crate) fn layout(&self) -> Layout {
+        self.layout
+    }
+}
+
+impl Drop for AlignedBuffer {
+    fn drop(&mut self) {
+        if !self.ptr.is_null() {
+            unsafe {
+                dealloc(self.ptr, self.layout);
+            }
+        }
+    }
+}
+
+// SAFETY: AlignedBuffer is Send because we control access to the raw pointer
+// and ensure proper ownership transfer
+unsafe impl Send for AlignedBuffer {}
+
+// NOTE: We do NOT implement Sync for AlignedBuffer because concurrent access
+// to the same buffer would be unsafe. Each buffer should be used by only one
+// thread at a time, or protected by explicit synchronization.
+
+thread_local! {
+    static OPT_BUFFER: std::cell::RefCell<Option<AlignedBuffer>> = std::cell::RefCell::new(None);
+    static UNOPT_BUFFER: std::cell::RefCell<Option<AlignedBuffer>> = std::cell::RefCell::new(None);
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct LiveVarsBuffer {
```

Comment:
I can only roughly guess what this struct is for. A doc string would help me a lot!